### ISSUE TITLE
[WebSocketClient] Open Handles after connectFailed?

### ISSUE THREAD

**Issue by alanraison** [theturtle32/WebSocket-Node#426]
I'm using a WebSocketClient in my integration tests, and I am testing the behaviour of the 'connectionFailed' event. From the examples (and the API) it seems that no clean-up is necessary after a connectionFailed event, yet jest is complaining that there are still open file handles at the end of the test. Perhaps this is due to an incorrect use of Promises, though I believe these are correct. 

Here is an example test:
```ts
async function connect(url: string): Promise<connection> {
  const client = new WebSocketClient();
  const promise = new Promise<connection>((resolve, reject) => {
    client.on('connectFailed', reject);
    client.on('connect', resolve);
  });
  client.connect(url);
  return promise;
}

test('not supplying a game id should fail', async () => {
  await expect(connect(`${wsUrl}?name=player1`)).rejects.toThrowError(
    /400 Bad Request/
  );
});
```

Running jest with `--detectOpenHandles` reports:
```
Jest has detected the following 5 open handles potentially keeping Jest from exiting:

  ●  TCPWRAP

      13 |     client.on('connect', resolve);
      14 |   });
    > 15 |   client.connect(url);
    
      at WebSocketClient.Object.<anonymous>.WebSocketClient.connect (../../../node_modules/websocket/lib/WebSocketClient.js:254:56)
```

I am using websocket 1.0.34.

Is this something I am doing wrong, or is jest just being pessimistic?

Thanks

**Comment by ibc** [comment]
Avoid jest and just paste your code into a file and run it with node as a script. Does the process exit or not? And does it exist automatically if you don't attempt to create any client WS connection?

**Comment by alanraison** [comment]
OK I have boiled it down to basic JS and it indeed does not exit:

```javascript
import ws from 'websocket';

const wsUrl = 'wss://websocket-returning-error';

async function connect(url) {
  const client = new ws.client();
  const promise = new Promise((resolve, reject) => {
    client.on('connectFailed', reject);
    client.on('connect', resolve);
  });
  client.connect(url);
  return promise;
}

try {
  await connect(`${wsUrl}?name=player1`);
} catch (e) {
  console.error(e);
}
```

**Comment by alanraison** [comment]
Interestingly, this *does* exit:

```javascript
import ws from 'websocket';

const wsUrl = 'wss://websocket-returning-error';

async function connect(url) {
  const client = new ws.client();
  const promise = new Promise((resolve, reject) => {
    client.on('httpResponse', reject);
    client.on('connectFailed', console.log);
    client.on('connect', resolve);
  });
  client.connect(url);
  return promise;
}

try {
  await connect(`${wsUrl}?name=player1`);
} catch (e) {
  console.error(e);
}
```
And the `console.log` of the `connectFailed` event doesn't get called but the `httpResponse` is printed in the catch block.

**Comment by ibc** [comment]
I don't know what to say. That's far from what I suggested.

**Comment by alanraison** [comment]
I'm not sure what you don't understand about what I did, but if you don't have anything helpful to say @ibc, just don't bother saying it.

I figured out how to deal with my connection-closure issue; it appears that my websocket server is sending an httpResponse, so I handle the `httpResponse` event, the payload for which contains a `destroy` method, which I call. Interestingly, this stops the `connectFailed` event from being fired, which is unexpected.

```javascript
const client = new ws.client();
client.on('connectFailed', console.error);
client.on('httpResponse', (r) => {
  console.error('httpResponse');
  r.destroy();
});
client.on('connect', (c) => {
  console.log('connected');
  c.close();
});
client.connect(`${wsUrl}?name=player`);
```

I hope this helps someone else use this library.
